### ISSUE TITLE
[suggestion] `scm` feature, for dirty builds to use local source

### ISSUE THREAD

**Issue by Mark-Hatch-Bose** [conan-io/conan#5126]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

```
Conan version 1.14.3
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.5 LTS
Release:	16.04
Codename:	xenial
```

See for context https://github.com/conan-io/conan/issues/4600

This is while using using the scm feature.
When local changes are made (source is not pristine), "last" commit revision is captured.
This is a problem when trying to rebuild dependencies accurately from source.
Source folder would be removed, and conan would checkout the "last" captured commit from scm. 
This commit does not reflect the local changes.
This change allows the use of `revision: "None"`, which will attempt to build using the local source folder.

Proposed changes at https://github.com/Mark-Hatch-Bose/conan.
Not opening a pull request for the changes until this is signaled as a desired feature.

Example usage:
```
def mark_dirty():
  git = tools.Git()
  if git.is_pristine():
    return "auto"
  else:
    return "None"

class TestRepo(ConanFile):
    scm = {
        "type": "git",
        "url": "git@github.com:Org/TestRepo.git",
        "revision": mark_dirty()
    }
```

**Comment by jgsogo** [comment]
Hi, @Mark-Hatch-Bose 

(Adding context and using my own words for your proposal) Some time ago we changed the behavior of that _SCM optimization_ to use local (potentially modified) sources only in the `create` command but not in the `install` one (https://github.com/conan-io/conan/issues/4218). Now you would want to request a special value for `scm::revision` to be able to use that _dangerous_ behavior again, right? Your proposal is safer than the previous behavior as it is something explicitly written by the user and not a default Conan behavior.

I'm afraid this could lead to unexpected behavior if Conan finds the `scm::revision="None"` but there are no sources in the cache or a recipe with that `None` is uploaded to a remote...

Maybe your use case would be best implemented through editable packages and building them in the local folders. If you have a package in editable mode and you request to build it as a dependency of another package it will be built in the local folder using local sources. This is something more likely to be in the roadmap. What do you think? Will it be a better way for your use case? Related issue: https://github.com/conan-io/conan/issues/4837

(Also related, to have these issues linked: https://github.com/conan-io/conan/issues/4732)

**Comment by Mark-Hatch-Bose** [comment]
> I'm afraid this could lead to unexpected behavior if Conan finds the scm::revision="None" but there are no sources in the cache or a recipe with that None is uploaded to a remote...

I would not consider it "dangerous" in this case since if you did not have a local source folder, it simply wouldn't build. SCM would try to checkout revision "None" and fail the build right? Like you said, also a user has to explicitly add "None", so it's a choice.

We would prevent "dirty" builds in Jenkins. All of our "officially" built packages should be pristine. Otherwise, they could be incorrect if there are modified files.
If the build fails and we see the error message saying it can't find revision "None" it seems that a user uploaded a package that was locally built, and it should be looked at and fixed/removed.

**The more important point is:**
It's much more dangerous that there is the perception that dependencies are built correctly, when it's not reflecting the actual source at build time of a dependency. If you built a package that has modified source, grabbing the "last" previous revision is incorrect and dangerous! 

Revision "None" would basically prevent this, as it forces your package builds to be pristine or fail to checkout. Not to mention it makes local workflows much easier. (Change your dependencies to your local builds, everything works and rebuilds at any build level)

> Maybe your use case would be best implemented through editable packages and building them in the local folders. If you have a package in editable mode and you request to build it as a dependency of another package it will be built in the local folder using local sources. This is something more likely to be in the roadmap. What do you think? Will it be a better way for your use case? Related issue: #4837

More than happy to switch to editable packages in the future when #4837 is implemented.
I will say that this proposed feature make the local build flow feel very "automatic" and seamless.
No need to explicitly put packages into `editable` mode.

**The workflow for multi-level builds:**
Go to A, modify it, build it, export.
Go to C modify it, build it, auto rebuilds B against local source A.
Go to B modify it, build it, export.
Go back to C, rebuild, everything is correct.
etc. etc.

When changes are ready, commit.
Webhook for build triggers. Pristine package is made.
At some point, the full tree is updated with new packages and rebuilt. If there is a revision "None", it'll fail to build upstream, and should be fixed. This shouldn't happen anyway, as Jenkins builds our repos for official packages, and is the only one who should be uploading them.

**Comment by jgsogo** [comment]
Yes, you are right, it is not _dangerous_, but Conan would be exporting a recipe that is invalid and, to be honest, I don't feel comfortable with it. Better than the `revision="None"` I would consider some `--keep-source` argument (some commands already have it) <- btw, I'm assuming this `--keep-source` is not working for your use case.

**Comment by Mark-Hatch-Bose** [comment]
> Conan would be exporting a recipe that is invalid and, to be honest, I don't feel comfortable with it.
Understandable, but you made a package with local changes, so is it really valid for `scm` to be gathering the last commit? The package made is storing a revision of source that does not match the revision stored in the `conanfile` `scm`. That seems very wrong.

I wasn't aware of `keep-source`, I just tested it , but it does not keep the source of downstream dependencies.

Test was:
A modified, exported
B modified, exported
C modified, exported

A modified, exported
Go to C, `-b outdated -ks`, grabs last revision of B rather than using local source of B with local changes.

**Comment by jgsogo** [comment]
I think here we have two main options, and probably both of them makes sense:
 1. package in _editable mode_ should implement the local build
 2. ensure the `-ks` flag behavior for all the packages built during a `conan install ... -ks` command.

**Comment by Mark-Hatch-Bose** [comment]
How does `-ks` work?
Would the upstream detect that `keep source` was used on a downstream dependency?

Also this means if someone uses `conan create` it wouldn't work right?
You'd have to `conan install`?

**Comment by jgsogo** [comment]
It looks like that `-ks` only work for the _first_ package, the one you are installing or creating. That it doesn't work for the packages being built as a consequence of `build=missing` or `build=outdated`. I dn't know if this is _by design_ or it is just the way it works.

**Comment by Mark-Hatch-Bose** [comment]
Yes, maybe a new flag is needed (to keep downstream source)

I tried using editable mode, but it is currently incompatible with my package naming scheme. It requires the reference of the package to match, but my current changes currently are generating a new version using a timestamp. I'll do some further testing.
```
mh70023@mh-AMBI-VM:~/.../remote-conan/TestRepo$ conan editable add . TestRepo/2.3.4@mh/testing
ERROR: Name and version from reference (TestRepo/2.3.4@mh/testing) and target conanfile.py (TestRepo/2.11.1-dirty.1558006190) must match
```

Even so I would want #4837 to be implemented.

From what I understand so far about editable workflows, it currently would be:
```
Test was:
Add A as editable
A modified, exported
Add B as editable
B modified, exported
Add C as editable
C modified, exported

A modified, exported
Go to B, rebuild. ( Until #4837 )
Go to C, rebuild.
```
Which isn't much different than my current flow with `dirty timestamp` as version.
Both require you to go to each build level of the tree and rebuild.

Anyway, I still really think that some users may want to mark packages with no `remote git commit` to be marked as such. It's very confusing that a package can be created with source that doesn't match the `revision` stored in the conanfile.

**Comment by jgsogo** [comment]
Let's recap the proposal and ask for some feedback about this feature request (ping @lasote):

The idea is to have a **special value `scm::revision="None"`** that will tell Conan to use the local sources optimization (the `scm_folder.txt` from SCM feature) not only in the `conan create`, but any time the recipe is built. The good thing about this is the `"None"` value is explicitly set by the user and it won't build if there are no local sources (CI systems, it will try to checkout the `None` revision)

**Comment by lasote** [comment]
So, I assume that eventually, you will export again the recipe to get the correct hash replacement, right? Shouldn't be more natural something like `scm::revision="local"` or so?

**Comment by Mark-Hatch-Bose** [comment]
@lasote Yes, I'm not tied to the name. To be similar to "auto" (special keyword) I think "local" or "none" is better.

Yes, once a commit is made, Jenkins webhook would be kicked off and official package would be made with the actual committed git revision acquired from `auto`.

This is only for easing local development. Making sure you are using the correct "local source" rather than last commit with "auto".

**Comment by lasote** [comment]
Merged. To be released at 1.20
